### PR TITLE
Add scylla-operator analyze command stub

### LINKED ISSUE #3: Create a scaffold for analyze command
We need to create a scaffold for the analyze command to begin with.

/kind feature
/priority important-soon

### PR DESCRIPTION
**Pull request by ttytek**
Add file analyze,go with NewAnalyzeCmd function with two flags (mode and dirPath) and simple flag value validation.

**Which issue is resolved by this Pull Request:**
Resolves #3

### CODE REVIEW

File: pkg/cmd/operator/analyze.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+func (o *AnalyzeOptions) Complete() error {
+	return nil
+}
+
+func (o *AnalyzeOptions) Run(streams genericclioptions.IOStreams, cmd *cobra.Command) error {
+	fmt.Printf("Analyze run with options: mode='%s', gatherDirPath='%s'\n", o.Mode, o.GatherDirPath)
+
+	// TODO
```
Comment (zimnx) [inline pkg/cmd/operator/analyze.go:41]: Lets not leave empty TODOs, it should have description what is to do. If there's a github issue, link should be there.

File: pkg/cmd/operator/analyze.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+			if err != nil {
+				return err
+			}
+
+			return nil
+		},
+	}
+
+	cmd.Flags().StringVarP(&o.Mode, "mode", "", o.Mode, "Specify the mode: live or gathered")
+	cmd.Flags().StringVarP(&o.GatherDirPath, "dirPath", "", o.GatherDirPath, "Specify the directory path for gathered data")
```
[the remark this answers is not in the document]
Comment (zimnx) [reply pkg/cmd/operator/analyze.go:82]: archive-path is good

File: pkg/cmd/operator/analyze.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+	"github.com/scylladb/scylla-operator/pkg/genericclioptions"
+	"k8s.io/apimachinery/pkg/util/errors"
+)
+
+type AnalyzeOptions struct{
+	Mode string
+	GatherDirPath string
+}
+
+func (o *AnalyzeOptions) Validate() error {
```
Comment (zimnx) [inline pkg/cmd/operator/analyze.go:16]: please move functions according to existing style. First Options struct definitions, then constructor (New*Options) then constructor for command and then Validate, Complete and Run at the end.

File: pkg/cmd/operator/analyze.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+			if err != nil {
+				return err
+			}
+
+			return nil
+		},
+	}
+
+	cmd.Flags().StringVarP(&o.Mode, "mode", "", o.Mode, "Specify the mode: live or gathered")
```
[the remark this answers is not in the document]
Comment (zimnx) [reply pkg/cmd/operator/analyze.go:81]: +1

File: pkg/cmd/operator/analyze.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+	}
+
+	o.AddFlags(cmd)
+
+	return cmd
+}
+
+func (o *AnalyzeOptions) AddFlags(cmd *cobra.Command) {
+	o.ClientConfig.AddFlags(cmd)
+	cmd.Flags().StringVarP(&o.ArchivePath, "archive-path", "", o.ArchivePath, "Path to must-gather archive directory")
```
Comment (zimnx) [inline pkg/cmd/operator/analyze.go:60]:
```suggestion
	o.ClientConfig.AddFlags(cmd)
	
	cmd.Flags().StringVarP(&o.ArchivePath, "archive-path", "", o.ArchivePath, "Path to must-gather archive directory")
```

Comment (zimnx) [inline pkg/cmd/operator/analyze.go:60]:
```suggestion
	o.ClientConfig.AddFlags(cmd)
	cmd.Flags().StringVarP(&o.ArchivePath, "archive-path", "", o.ArchivePath, "Path to compressed must-gather  archive or directory having must-gather structure")
```

File: pkg/cmd/operator/analyze.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+	if len(o.Kubeconfig) == 0 && len(o.ArchivePath) == 0 {
+		errs = append(errs, fmt.Errorf("kubeconfig and archive-path can't both be empty"))
+	}
+
+	return errors.NewAggregate(errs)
+}
+
+func (o *AnalyzeOptions) Complete() error {
+
```
Comment (zimnx) [inline pkg/cmd/operator/analyze.go:85]:
```suggestion
```

File: pkg/cmd/operator/analyze.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+func NewAnalyzeCmd(streams genericclioptions.IOStreams) *cobra.Command {
+	o := NewAnalyzeOptions(streams)
+
+	cmd := &cobra.Command{
+		Use:   "analyze",
+		Short: "Run scylla-operator analyze.",
+		Long:  "Run the scylla anallyze to analyze either live cluster or gathered data",
+		RunE: func(cmd *cobra.Command, args []string) error {
+
```
Comment (zimnx) [inline pkg/cmd/operator/analyze.go:33]:
```suggestion
```

File: pkg/cmd/operator/analyze.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+func NewAnalyzeOptions(streams genericclioptions.IOStreams) *AnalyzeOptions {
+	return &AnalyzeOptions{}
+}
+
+func NewAnalyzeCmd(streams genericclioptions.IOStreams) *cobra.Command {
+	o := NewAnalyzeOptions(streams)
+
+	cmd := &cobra.Command{
+		Use: "analyze",
+		Short: "Run the scylla analyze.",
```
[the remark this answers is not in the document]
Comment (ttytek) [reply pkg/cmd/operator/analyze.go]: done

File: pkg/cmd/operator/analyze.go [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+		Long: "Run the scylla anallyze to analyze either live cluster or gathered data",
```
[the remark this answers is not in the document]
Comment (ttytek) [reply pkg/cmd/operator/analyze.go]: done

File: pkg/cmd/operator/analyze.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ttytek) [reply pkg/cmd/operator/analyze.go]: done

File: pkg/cmd/operator/analyze.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ttytek) [reply pkg/cmd/operator/analyze.go]: done

File: pkg/cmd/operator/analyze.go [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+	return errors.NewAggregate(errs)
+}
+
+func (o *AnalyzeOptions) Complete() error {
+	return nil
+}
+
+func (o *AnalyzeOptions) Run(streams genericclioptions.IOStreams, cmd *cobra.Command) error {
+	fmt.Printf("Analyze run with options: mode='%s', gatherDirPath='%s'\n", o.Mode, o.GatherDirPath)
```
[the remark this answers is not in the document]
Comment (ttytek) [reply pkg/cmd/operator/analyze.go]: done

File: pkg/cmd/operator/analyze.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ttytek) [reply pkg/cmd/operator/analyze.go]: done

File: pkg/cmd/operator/analyze.go [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+			return nil
+		},
+	}
+
+	cmd.Flags().StringVarP(&o.Mode, "mode", "", o.Mode, "Specify the mode: live or gathered")
+	cmd.Flags().StringVarP(&o.GatherDirPath, "dirPath", "", o.GatherDirPath, "Specify the directory path for gathered data")
+
+	return cmd
+}
```
[the remark this answers is not in the document]
Comment (ttytek) [reply pkg/cmd/operator/analyze.go]: done

File: pkg/cmd/operator/analyze.go [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+	o := NewAnalyzeOptions(streams)
+
+	cmd := &cobra.Command{
+		Use: "analyze",
+		Short: "Run the scylla analyze.",
+		Long: "Run the scylla anallyze to analyze either live cluster or gathered data",
+		RunE: func(cmd *cobra.Command, args []string) error{
+
+			err := o.Validate()
+
```
[the remark this answers is not in the document]
Comment (ttytek) [reply pkg/cmd/operator/analyze.go]: done

File: pkg/cmd/operator/analyze.go (hunk 5 above)
Comment (ttytek) [reply pkg/cmd/operator/analyze.go]: done

Comment (ttytek) [reply pkg/cmd/operator/analyze.go]: done

File: pkg/cmd/operator/analyze.go (hunk 6 above)
Comment (ttytek) [reply pkg/cmd/operator/analyze.go]: done

File: pkg/cmd/operator/analyze.go (hunk 7 above)
Comment (ttytek) [reply pkg/cmd/operator/analyze.go]: done

**Review by zimnx — APPROVED** [review]
Thanks for updates. PR is missing kind and priority labels, please add them.

/assign rzetelskik

File: pkg/cmd/operator/analyze.go [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+	o.AddFlags(cmd)
+
+	return cmd
+}
+
+func (o *AnalyzeOptions) AddFlags(cmd *cobra.Command) {
+	o.ClientConfig.AddFlags(cmd)
+
+	cmd.Flags().StringVarP(&o.ArchivePath, "archive-path", "", o.ArchivePath, "Path to compresed must-gather archive or directory having must-gather structure")
```
Comment (rzetelskik) [inline pkg/cmd/operator/analyze.go:60]:
```suggestion
	cmd.Flags().StringVarP(&o.ArchivePath, "archive-path", "", o.ArchivePath, "Path to a compressed must-gather archive or a directory having must-gather structure.")
```

File: pkg/cmd/operator/analyze.go [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+	"github.com/scylladb/scylla-operator/pkg/genericclioptions"
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/util/errors"
+	cliflag "k8s.io/component-base/cli/flag"
+	"k8s.io/klog/v2"
+)
+
+type AnalyzeOptions struct {
+	genericclioptions.ClientConfig
+	ArchivePath string
```
Comment (rzetelskik) [inline pkg/cmd/operator/analyze.go:16]:
```suggestion
	genericclioptions.ClientConfig
	
	ArchivePath string
```

let's separate embeddings from proprietary flags

File: pkg/cmd/operator/analyze.go [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+	err := o.ClientConfig.Complete()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (o *AnalyzeOptions) Run(streams genericclioptions.IOStreams, cmd *cobra.Command) error {
+	klog.Infof("%s run", cmd.Name())
```
Comment (rzetelskik) [inline pkg/cmd/operator/analyze.go:94]: why omit the version?

File: pkg/cmd/operator/analyze.go [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+	if err != nil {
+		errs = append(errs, fmt.Errorf("archive-path is not a valid filepath"))
+	}
+
+	if len(o.Kubeconfig) != 0 && len(o.ArchivePath) != 0 {
+		errs = append(errs, fmt.Errorf("kubeconfig and archive-path can't both be set"))
+	}
+
+	if len(o.Kubeconfig) == 0 && len(o.ArchivePath) == 0 {
+		errs = append(errs, fmt.Errorf("kubeconfig and archive-path can't both be empty"))
```
Comment (rzetelskik) [inline pkg/cmd/operator/analyze.go:78]: empty kubeconfig is fine, a default is used in that case. If both are empty we implicitly assume working on a live cluster.

File: pkg/cmd/operator/analyze.go [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+
+func (o *AnalyzeOptions) Validate() error {
+	var errs []error
+
+	errs = append(errs, o.ClientConfig.Validate())
+
+	_, err := os.Stat(o.ArchivePath)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("archive-path is not a valid filepath"))
+	}
```
Comment (rzetelskik) [inline pkg/cmd/operator/analyze.go:71]: 1. what if `o.ArchivePath` is an empty string (i.e. you want to use kubeconfig)?
2. an err here does not have to indicate an invalid filepath

File: pkg/cmd/operator/analyze.go [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+)
+
+type AnalyzeOptions struct {
+	genericclioptions.ClientConfig
+	ArchivePath string
+}
+
+func NewAnalyzeOptions(streams genericclioptions.IOStreams) *AnalyzeOptions {
+	return &AnalyzeOptions{
+		ClientConfig: genericclioptions.NewClientConfig("scylla-analyze"),
```
[the remark this answers is not in the document]
Comment (ttytek) [reply pkg/cmd/operator/analyze.go]: done

File: pkg/cmd/operator/analyze.go (hunk 13 above)
Comment (ttytek) [reply pkg/cmd/operator/analyze.go]: done

File: pkg/cmd/operator/analyze.go (hunk 14 above)
Comment (ttytek) [reply pkg/cmd/operator/analyze.go]: done

File: pkg/cmd/operator/analyze.go (hunk 15 above)
Comment (ttytek) [reply pkg/cmd/operator/analyze.go]: fixed

File: pkg/cmd/operator/analyze.go (hunk 16 above)
Comment (ttytek) [reply pkg/cmd/operator/analyze.go]: removed this check

File: pkg/cmd/operator/analyze.go (hunk 17 above)
Comment (ttytek) [reply pkg/cmd/operator/analyze.go]: Fixed error condition

File: pkg/cmd/operator/analyze.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (rzetelskik) [reply pkg/cmd/operator/analyze.go]: (this comment still applies)

[the remark this answers is not in the document]
Comment (rzetelskik) [reply pkg/cmd/operator/analyze.go]: I like it, but I think we can make it a bit more concise and highlight that it only focuses on resources related to Scylla Operator. I also think the second line is redundant - flags description should already explain it. As for the command not being implemented yet, I think it'd be more future-proof to say it's experimental (like must-gather does):

How about:
`analyze traverses resources associated with Scylla Operator and diagnoses common issues.

This command is experimental and subject to change without notice.
`

File: pkg/cmd/operator/analyze.go (hunk 17 above)
Comment (ttytek) [reply pkg/cmd/operator/analyze.go]: Fixed

File: pkg/cmd/operator/analyze.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (ttytek) [reply pkg/cmd/operator/analyze.go]: Done

**Review by rzetelskik — APPROVED** [review]
/lgtm
thanks and congrats on the first merged PR @ttytek :tada:

**Pull request merged**